http-clients/java-nio-client/src/test/java/software/amazon/awssdk/http/nio/java/JavaHttpClientNioAsyncHttpClientWireMockTest.java

```diff
@@ -0,0 +1,302 @@
+                .build()).join();
+
+        recorder.completeFuture.get(2, TimeUnit.SECONDS);
+
+        assertThat(wiremockTrafficListener.response.toString().endsWith(content));
+    }
+
+    private SdkHttpFullRequest createRequest(URI uri) {
+        return createRequest(uri, "/", "", SdkHttpMethod.GET, emptyMap(), emptyMap());
+    }
+
+    private SdkHttpFullRequest createRequest(URI uri,
+                                             String resourcePath,
+                                             String body,
+                                             SdkHttpMethod method,
+                                             Map<String, String> params,
+                                             Map<String, List<String>> headers) {
+        String contentLength = body == null ? null : String.valueOf(body.getBytes(UTF_8).length);
+        return SdkHttpFullRequest.builder()
+                .uri(uri)
+                .method(method)
+                .encodedPath(resourcePath)
+                .applyMutation(b -> params.forEach(b::putRawQueryParameter))
+                .applyMutation(b -> {
+                    b.putHeader("Host", uri.getHost());
+                    if (contentLength != null) {
+                        b.putHeader("Content-Length", contentLength);
+                    }
+                })
+                .applyMutation(b -> headers.forEach(b::putHeader))
+                .build();
+    }
+
+
+    private SdkHttpContentPublisher createProvider(String body) {
+
+        return new SdkHttpContentPublisher() {
+            Flowable<ByteBuffer> flowable = Flowable.just(ByteBuffer.wrap(body.getBytes()));
+            @Override
+            public Optional<Long> contentLength() {
+                return Optional.of((long) body.length());
+            }
+
+            @Override
+            public void subscribe(Subscriber<? super ByteBuffer> s) {
+                flowable.subscribeWith(s);
+            }
+        };
+    }
+
+
+    @Test
+    public void testConnectionTimeoutError() throws Exception{
+//        String expectedErrorMsg = "java.net.http.HttpConnectTimeoutException: HTTP connect timed out";
+
+        try {
+            System.setProperty("jdk.internal.httpclient.debug", "true");
+            SdkAsyncHttpClient customClient = JavaHttpClientNioAsyncHttpClient.builder()
+                    .connectionTimeout(Duration.ofMillis(1))
+                    .build();
+
+            List<CompletableFuture<Void>> futures = new ArrayList<>();
+            for (int i = 0; i < 2; i++) {
+                futures.add(makeSimpleRequestAndReturnResponseHandler(customClient).completeFuture);
+            }
+
+//            assertThatThrownBy(() -> CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join())
+//                    .hasMessageContaining(expectedErrorMsg);
+
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    @Test
+    public void protocolChangedTest() {
```

Comment:
Since we didn't reserve an interface here so I can't get access to the HttpClient created in JavaHttpClientNioAsyncHttpClient, so I could only use the debug mode to check whether the protocol was set as what was in the protocol() method. Here I'm not sure how to check the protocol rather than this way or whether we need this test here.

Reply:
Let's discuss this in person. There's most likely an easier way to test this.

Reply:
Sorry Dongie, I forgot the real reason why I didn't add an interface to get the HttpClient in JavaHttpClientNioAsyncHttpClient class when I discussed this with you, it's because the object created by the builder is SdkAsyncHttpClient, and there is no such an interface in that class, so I didn't find a way to get the HttpClient.

Reply:
Should be fine to cast it back to a `JavaHttpClientNioAsyncHttpClient` since we know it is an instance of that class.